[Distributed] Fix @parametrize on unordered iterable in distributed test (again)

eqy opened the pull request:
Same fix as https://github.com/pytorch/pytorch/pull/159793, as this was broken again in https://github.com/pytorch/pytorch/pull/161476, assuming to be due to a botched rebase

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @mruberry

eqy committed:
Change state dict implementation to a tuple

Skylion007 commented on test/distributed/fsdp/test_distributed_checkpoint.py:
```diff
@@ -31,10 +31,10 @@
 
 
-_DISTRIBUTED_STATE_DICT_IMPLS = {
+_DISTRIBUTED_STATE_DICT_IMPLS = (
```
Would be better to use OrderedSet here for clarity?

eqy committed:
Add note on iterable order for distributed tests

Added a note about the importance of order in the iterable for distributed tests.
